### DISCUSSION TITLE
当并发创建会话时，EngineRef判断当前引擎数量为空时，会不会导致并发创建引擎？

### DISCUSSION THREAD

**Discussion by shenbinglife** [apache/kyuubi#1952] [Q&A] [answered]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

当并发创建会话时，EngineRef判断当前引擎数量为空时，会不会导致并发创建引擎？

 getServerHost(zkClient, engineSpace).getOrElse {
        create(zkClient, extraEngineLog)
      }

这段代码在并发上是不是有问题？ 

### Affects Version(s)

1.5.0

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

**Comment by zhouyifan279** [comment, marked as answer]
不会。`org.apache.kyuubi.engine.EngineRef#create`中会获取分布式锁。

PS：不确定的问题可以先在[Disscutions](https://github.com/apache/incubator-kyuubi/discussions)中创建问题。Issue中不适合做FAQ。

**Reply by shenbinglife** [reply]
但是，create 中分布式锁方法只能保证创建引擎的进程是唯一的，不能保证重复创建引擎。
例如用户A 同时创建会话S1 和S2，
S1 和 S2 获取引擎时， 会同时进入getServerHost().getOrElse { create () } 方法时，由于该方法不是同步的，所以可以同时进入create() 方法。

由于Zk分布式锁的原因，假设S1 进入了Create 方法， S2 被阻塞； 但是当S1 创建完成引擎，释放分布式锁后， S2 会话会从create方法的分布式锁处唤醒继续执行创建引擎。这就创建了2个引擎

我认为，需要在create方法中增加获取当前引擎的方法

**Reply by shenbinglife** [reply]
或者getOrCreate 方法设为synchronized

**Reply by ulysses-you** [reply]
@shenbinglife  are you looking at the code of master branch ? It seems correct to me
```scala
org.apache.kyuubi.engine.EngineRef

private def create(
    zkClient: CuratorFramework,
    extraEngineLog: Option[OperationLog]): (String, Int) = tryWithLock(zkClient) {
// Get the engine address ahead if another process has succeeded
var engineRef = getServerHost(zkClient, engineSpace)
if (engineRef.nonEmpty) return engineRef.get

```

The code requires the zk lock first, then check if there exists an active engine.

**Reply by shenbinglife** [reply]
very sorry，我的代码是被二次开发出现的问题，删除了关键的代码，master版本非常正确

**Reply by ulysses-you** [reply]
@shenbinglife  thank you for your confirming,  welcome to Kyuubi !
